processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java

```diff
@@ -289,16 +298,99 @@ public Row apply(Row input)
   }
 
   @Override
+  public Function<Row, Row> makePostComputeManipulatorFn(
+      final GroupByQuery query,
+      final MetricManipulationFn fn
+  )
+  {
+    final Set<String> optimizedDims = optimizedExtractionDims(query, false);
+    final Function<Row, Row> preCompute = makePreComputeManipulatorFn(query, fn);
+    if (optimizedDims.isEmpty()) {
+      return preCompute;
+    }
+
+    // If we have optimizations that can be done at this level, we apply them here
+
+    final Map<String, ExtractionFn> extractionFnMap = new HashMap<>();
+    for (DimensionSpec dimensionSpec : query.getDimensions()) {
+      final String dimension = dimensionSpec.getOutputName();
+      if (optimizedDims.contains(dimension)) {
+        extractionFnMap.put(dimension, dimensionSpec.getExtractionFn());
+      }
+    }
+
+    return new Function<Row, Row>()
+    {
+      @Nullable
+      @Override
+      public Row apply(Row input)
+      {
+        Row preRow = preCompute.apply(input);
+        if (preRow instanceof MapBasedRow) {
+          MapBasedRow preMapRow = (MapBasedRow) preRow;
+          Map<String, Object> event = Maps.newHashMap(preMapRow.getEvent());
+          for (String dim : optimizedDims) {
+            final Object eventVal = event.get(dim);
+            event.remove(dim);
+            event.put(dim, extractionFnMap.get(dim).apply(eventVal == null ? "" : eventVal));
+          }
+          return new MapBasedRow(preMapRow.getTimestamp(), event);
+        } else {
+          return preRow;
+        }
+      }
+    };
+  }
+
+  @Override
   public TypeReference<Row> getResultTypeReference()
   {
     return TYPE_REFERENCE;
   }
 
   @Override
-  public QueryRunner<Row> preMergeQueryDecoration(QueryRunner<Row> runner)
+  public QueryRunner<Row> preMergeQueryDecoration(final QueryRunner<Row> runner)
```

Comment:
Btw, if it's not possible to move this to `mergeResults` for some reason, I'd love to know that reason so that we can update the docs with why this method exists.